modules/conversation_cache.py

```diff
@@ -0,0 +1,38 @@
+from collections import OrderedDict
+
+
+# TODO
+# this is not thread safe
+# this does not scale
+# this does not work in a distributed system
+# this is a placeholder to be replaced with something like memcached
+class LRUCache:
+    def __init__(self, max_size):
+        self.cache = OrderedDict()
+        self.max_size = max_size
+
+
+    # Get the value from the cache.
+    # Returns the value if the key exists, empty string if not
+    def get(self, key):
+        if key in self.cache:
+            # Move the accessed key to the end to mark it as most recently used
+            self.cache.move_to_end(key)
+            return self.cache[key]
+        else:
+            return ""
+
+    # Adds a new key-value pair to the cache.  If the key already exists the new value will be appended to the existing value
+    def upsert(self, key, value):
+        oldVal=self.get(key)
+        if oldVal:
+            # TODO limit length of cached value
+            self.cache[key]=oldVal+"\n"+value
```

Comment:
This operation looks a bit strange to when a key is already present. I hope it will be refined later.

Reply:
the primary client usage currently is to append the new LLM response to the previous conversation history, hence why it behaves this way.

there's certainly no reason it couldn't be a more traditional cache in which the client is responsible for defining the new value and updating the key explicitly.